### PR TITLE
Feature/can visit ingress app

### LINKED ISSUE #52: Automate check `checkCanVisitIngressApp`
Currently, the check `checkCanVisitIngressApp` asks the user to execute manual steps. Implement an automatic version of this check similar to the other port forwarding checks.

### PR DESCRIPTION
**Pull request by mkretz**
I have implemented the check whether the deployment can be reached from the ingress controller pods. This involves 

- getting the relevant ingress resources for the deployment's pods
- determining the ingress controllers responsible for those ingresses
- port forwarding to the pods of the ingress controlles
- making an HTTP request to the forwarded ports 

Resolves #52

### CODE REVIEW

File: internal/livelint/run_checks.go [hunk 1]
```diff
@@ -302,6 +307,24 @@ func (n *Livelint) RunChecks(namespace, deploymentName string, isVerbose bool) e
...
 		}
 
+		result = checkHasValidIngressClass(ingress, ingressClasses)
+		statusMsg.AddCheckResult(result)
+		n.ui.Send(statusMsg)
+		if result.HasFailed {
+			return nil
+		}
+
+		result, err = n.checkCanAccessAppFromIngressController(ingress, ingressClasses)
```
Comment (cloudlena) [inline internal/livelint/run_checks.go:317]: Why do we diverge from the flow chart here?

**Review by cloudlena — CHANGES_REQUESTED** [review]

File: internal/livelint/check_can_access_app.go [hunk 2]
```diff
@@ -43,14 +43,15 @@ func (n *Livelint) checkCanAccessApp(pods []apiv1.Pod) CheckResult {
 	}
 }
 
-func (n *Livelint) canPortForward(pod apiv1.Pod, port int32) bool {
+func (n *Livelint) canPortForward(pod apiv1.Pod, port int32, check func(uint16) bool) bool {
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/check_can_access_app.go:46]: Agreed. Done.

File: internal/livelint/check_has_valid_ingressclass.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package livelint
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/check_has_valid_ingressclass.go:1]: Agreed. Done.

File: .gitignore [hunk 4]
```diff
@@ -14,3 +14,6 @@ bin/
 
 # Goreleaser Output
 dist/
+
+# Logfiles written during testing
```
[the remark this answers is not in the document]
Comment (mkretz) [reply .gitignore:18]: Jep, removed.

File: internal/livelint/check_has_valid_ingressclass.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"fmt"
+
+	netv1 "k8s.io/api/networking/v1"
+)
+
+func checkHasValidIngressClass(ingress netv1.Ingress, ingressClasses map[string]netv1.IngressClass) CheckResult {
+	result := CheckResult{HasFailed: false, Message: fmt.Sprintf("Ingress %s has a valid ingress class", ingress.Name)}
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/check_has_valid_ingressclass.go:10]: Agreed. Done.

File: internal/livelint/get_ingressclasses.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package livelint
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/get_ingressclasses.go:1]: Done.

File: internal/livelint/get_ingresscontroller.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+func isControllerPod(pod apiv1.Pod, controllerName string) bool {
+	found := false
+	for _, container := range pod.Spec.Containers {
+		for _, arg := range container.Args {
+			if strings.Contains(arg, controllerName) {
+				found = true
+				break
+			}
+		}
+		if !found {
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/get_ingresscontroller.go:60]: Duh! Absolutely. Done.

File: internal/livelint/get_ingresscontroller.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	for _, ownerRef := range pod.OwnerReferences {
+		if *ownerRef.Controller {
+			switch ownerRef.Kind {
+			case "ReplicaSet":
+				replicaSet, err := n.k8s.AppsV1().ReplicaSets(pod.Namespace).Get(context.Background(), ownerRef.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, fmt.Errorf("error getting owning ReplicaSet's label selector for pod %s: %w", pod.Name, err)
+				}
+				labelSelector = replicaSet.Spec.Selector.MatchLabels
+				found = true
```
[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/get_ingresscontroller.go:89]: Done, as far as I can see.

File: main.go [hunk 9]
```diff
@@ -84,6 +91,11 @@ func main() {
 
 					ll.Start()
 
+					runChecksErr := ll.RunChecks(c.String("namespace"), args.Get(0), c.Bool("verbose"))
```
[the remark this answers is not in the document]
Comment (mkretz) [reply main.go:94]: Jep. Done.

File: internal/livelint/run_checks.go (hunk 1 above)
Comment (mkretz) [reply internal/livelint/run_checks.go:317]: There are two points where we diverge:

1. additional check `checkHasValidIngressClass`: this is an opportunity to report an additional potential error. This only comes up when you are trying to determine the ingress controller pods which are responsible for a particular `Ingress` resource via the API.  I have adapted the flow chart for this.
2. failure branch after `checkCanSeeBackends`: this is simply not implemented yet. I have opened #71 for this.

[the remark this answers is not in the document]
Comment (mkretz) [reply internal/livelint/run_checks.go:317]: Agreed. Done.

**Pull request merged**